### PR TITLE
07 memory management module

### PR DESCRIPTION
**Pull request by DomninDmitry-dev**
Homework MemoryManagement test module.
Files test in directory: /sys/my_memory_management/
- kmalloc-test
- kzalloc-test
- vmalloc-test
- get_free_pages-test

This module testing:
functions: kmalloc, kzalloc, vmalloc, get_free_pages;
measure time of each allocation/freeing.

Build this module:
make SYSTEM=qemu
Clean this module:
make clean

### CODE REVIEW

File: 07-MemoryManagement/module/result/get_free_pages.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+2^X	Buffer size(bytes)	Allocation time(nS)	Freeing time(nS)
+2^0+0              1		001596			001056
+2^0+1              2		000427			000637
+2^1+0              2		000345			000260
+2^1+1              3		000404			000275
+2^2+0              4		000356			000268
+2^2+1              5		000533			000460
+2^3+0              8		001844			001054
+2^3+1              9		003240			001375
+2^4+0             16		Mem error
```
Comment (yekovalyov) [inline 07-MemoryManagement/module/result/get_free_pages.txt]: What is the maximum amount of memory allocated by __get_free_pages?

File: 07-MemoryManagement/module/mymodule.c [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		if (strcmp(attr->name, KMALLOC_FILE_NAME) == 0)
+			p = kmalloc(allocsize, GFP_KERNEL);
+		else if (strcmp(attr->name, KZALLOC_FILE_NAME) == 0)
+			p = kzalloc(allocsize, GFP_KERNEL);
+		else if (strcmp(attr->name, VMALLOC_FILE_NAME) == 0)
+			p = vmalloc(allocsize);
+		else if (strcmp(attr->name, GETFREEPAGES_FILE_NAME) == 0) {
+			if (j == 0) {
+				p = (char *)__get_free_pages(GFP_KERNEL, allocsize);
+				allocsize *= PAGE_SIZE;
```
Comment (yekovalyov) [inline 07-MemoryManagement/module/mymodule.c]: Please check allocsize calculation correctness for __get_free_pages.
>		allocsize = my_pow(2, i) + j;
>		getrawmonotonic(&ts);
>		j1 = SEC_TO_NS(ts.tv_sec) + ts.tv_nsec;
>		if (strcmp(attr->name, KMALLOC_FILE_NAME) == 0)
>			p = kmalloc(allocsize, GFP_KERNEL);
>		else if (strcmp(attr->name, KZALLOC_FILE_NAME) == 0)
>			p = kzalloc(allocsize, GFP_KERNEL);
>		else if (strcmp(attr->name, VMALLOC_FILE_NAME) == 0)
>			p = vmalloc(allocsize);
>		else if (strcmp(attr->name, GETFREEPAGES_FILE_NAME) == 0) {
>			if (j == 0) {
>				p = (char *)__get_free_pages(GFP_KERNEL, allocsize);
>				allocsize *= PAGE_SIZE;

File: 07-MemoryManagement/module/result/get_free_pages.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+$ cat /sys/my_memory_management/get_free_pages-test 
+Type mem: get_free_pages-test
+2^X	Buffer size(bytes)	Allocation time(nS)	Freeing time(nS)
+2^0+0           4096		003718			000959
+2^1+0           8192		001152			000822
+2^2+0          16384		001546			000659
+2^3+0          32768		003551			001724
+2^4+0          65536		Mem error
```
Comment (yekovalyov) [inline 07-MemoryManagement/module/result/get_free_pages.txt]: What is the maximum size of memory allocated by get_free_pages?

**Comment by DomninDmitry-dev** [comment]
Maximum size of memory - 32768 bytes.
I don't understand what the problem.

**Comment by DomninDmitry-dev** [comment]
Fixed.

File: 07-MemoryManagement/module/mymodule.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	for (i = 0, j = 0; i < 64; j++) {
+		if (j == 2) {
+			i++;
+			j = 0;
+		}
+
+		if (strcmp(attr->name, GETFREEPAGES_FILE_NAME) == 0)
+			allocsize = i;
+		else
+			allocsize = my_pow(2, i) + j;
```
Comment (yekovalyov) [inline 07-MemoryManagement/module/mymodule.c]: What is the reason for using pow function for base 2? It can be implemented using shifting to left.

**Comment by DomninDmitry-dev** [comment]
What is the reason for using pow function for base 2? It can be implemented using shifting to left.
Done, fixed.

**Pull request merged**